lldb/source/Plugins/LanguageRuntime/Swift/SwiftLanguageRuntime.cpp

```diff
@@ -2584,37 +2588,142 @@ lldb::addr_t SwiftLanguageRuntime::GetAsyncContext(RegisterContext *regctx) {
+  case ValueType::isRegisterPlusOffset: {
+    unsigned regnum = cfa_loc.GetRegisterNumber();
+    if (llvm::Expected<addr_t> regvalue =
+            ReadRegisterAsAddress(regctx, regkind, regnum))
+      return *regvalue + cfa_loc.GetOffset();
+    else
+      return regvalue;
+  }
+  case ValueType::isConstant:
+  case ValueType::isDWARFExpression:
+  case ValueType::isRaSearch:
+  case ValueType::isRegisterDereferenced:
+  case ValueType::unspecified:
+    break;
+  }
+  return llvm::createStringError(
+      "SwiftLanguageRuntime: Unsupported FA location type = %d",
+      cfa_loc.GetValueType());
+}
+
+/// Attempts to use UnwindPlans that inspect assembly to recover the entry value
+/// of the async context register. This is a simplified version of the methods
+/// in RegisterContextUnwind, since plumbing access to those here would be
+/// challenging.
+static llvm::Expected<addr_t> ReadAsyncContextRegisterFromUnwind(
+    SymbolContext &sc, Process &process, Address pc, Address func_start_addr,
+    RegisterContext &regctx, AsyncUnwindRegisterNumbers regnums) {
+  FuncUnwindersSP unwinders =
+      pc.GetModule()->GetUnwindTable().GetFuncUnwindersContainingAddress(pc,
+                                                                         sc);
+  if (!unwinders)
+    return llvm::createStringError("SwiftLanguageRuntime: Failed to find "
+                                   "function unwinder at address 0x%8.8" PRIx64,
+                                   pc.GetFileAddress());
 
-  const uint8_t *expr = triple == llvm::Triple::x86_64
-                            ? g_cfa_dwarf_expression_x86_64
-                            : g_cfa_dwarf_expression_arm64;
-  auto size = triple == llvm::Triple::x86_64
-                  ? sizeof(g_cfa_dwarf_expression_x86_64)
-                  : sizeof(g_cfa_dwarf_expression_arm64);
-  if (with_double_deref)
-    return llvm::ArrayRef<uint8_t>(expr, size);
-  return llvm::ArrayRef<uint8_t>(expr, size - 1);
+  Target &target = process.GetTarget();
+  UnwindPlanSP unwind_plan =
+      unwinders->GetUnwindPlanAtNonCallSite(target, regctx.GetThread());
+  if (!unwind_plan)
+    return llvm::createStringError(
+        "SwiftLanguageRuntime: Failed to find non call site unwind plan at "
+        "address 0x%8.8" PRIx64,
+        pc.GetFileAddress());
+
+  const RegisterKind unwind_regkind = unwind_plan->GetRegisterKind();
+  UnwindPlan::RowSP row = unwind_plan->GetRowForFunctionOffset(
+      pc.GetFileAddress() - func_start_addr.GetFileAddress());
+
+  // To request info about a register from the unwind plan, the register must
+  // be in the same domain as the unwind plan's registers.
+  uint32_t async_reg_unwind_regdomain;
+  if (!regctx.ConvertBetweenRegisterKinds(
+          regnums.GetRegisterKind(), regnums.async_ctx_regnum, unwind_regkind,
+          async_reg_unwind_regdomain))
+    return llvm::createStringError(
+        "SwiftLanguageRuntime: Failed to convert register domains");
```

Comment:
Could this be an assert? I'm assuming, perhaps wrongly, that this should never happen. I think the log messages should be stuff that could happen for understood reasons.

Reply:
Honestly I am not sure when a register domain conversion can occur.
But let's say that it does: if we replace the `return` of an error with an assert, then a production compiler (no asserts) will keep going and do very very wrong things in the rest of this function (because we are also not returning anymore?).  This doesn't seem better than simply logging and letting the traditional unwinder take over.

Reply:
we can have both: assert in assert builds, and return with an error in production. The better the test coverage, the  more more useful an assert is at catching issues before they make it to production.

Also asserts highlight to readers and to authors of future changes, the nature of the error. Is it a coding/build/confitguration error, or is it a legitimate kind of runtime error.

Reply:
Yeah I don't mind having an assert that is only hit in DEBUG builds in CI or on an lldb developer's desktop.  asserts in production/release builds are much trickier IMO.

Reply:
Sounds good! I added an unreachable statement prior to returning the error. This way we get a hard crash in debug builds and, in release builds, we gracefully fallback to other unwind plans